types/queries.go

```diff
@@ -23,16 +23,23 @@ func RustQuery(querier Querier, binRequest []byte, gasLimit uint64) QuerierResul
 	var request QueryRequest
 	err := json.Unmarshal(binRequest, &request)
 	if err != nil {
-		return ToQuerierResult(nil, UnsupportedRequest{err.Error()})
+		return QuerierResult{
+			Err: &SystemError{
+				InvalidRequest: &InvalidRequest{
+					Err:     err.Error(),
+					Request: binRequest,
+				},
+			},
+		}
 	}
 	bz, err := querier.Query(request, gasLimit)
 	return ToQuerierResult(bz, err)
 }
 
 // This is a 2-level result
 type QuerierResult struct {
-	Ok  *QueryResponse `json:"Ok,omitempty"`
-	Err *SystemError   `json:"Err,omitempty"`
+	Ok  *QueryResponse `json:"ok,omitempty"`
+	Err *SystemError   `json:"error,omitempty"`
```

Comment:
What is the current state of SystemError? And the `QuerierResult` returned from the wasm import? This is still a typed struct?

I need to look at the cosmwasm implementation again (after finishing my pending deliverables), but maybe we can simplify this? Or is that something in 0.12 when you refactor StdError/SystemErr types

Reply:
System error still exists as a structured error:

```rust
pub enum SystemResult<S> {
    Ok(S),
    #[serde(rename = "error")]
    Err(SystemError),
}

// …

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
#[non_exhaustive]
pub enum SystemError {
    InvalidRequest { error: String, request: Binary },
    InvalidResponse { error: String, response: Binary },
    NoSuchContract { addr: HumanAddr },
    Unknown {},
    UnsupportedRequest { kind: String },
}
```

What I struggle with is a clear path to detect in which category a querier error falls.

Reply:
Ah... yeah, we end up with something like: `SystemResult<Result<T, String>>` right?

Reply:
Maybe we collapse system result into a string as well and queriers just return `Result<T, String>`. I have never once seen someone use those detailed messages, although it did seem like a good idea at the time.

Reply:
I think it is valuable to distinguish where the error comes from (system or contract internal). We could do `Result<Result<T, String>, String>` but as long as we have a fixed number of simple error cases, I don't see harm in the enum.